File: daily_problems/2024/06/0615/personal_submission/cf811d_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3")
using namespace std;

pair<int, int> op(char c) {
    cout << c << '\n';
    cout << flush;
    int x, y;
    cin >> x >> y;
    x --, y --;
    return {x, y};
}

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;
    
    vector<string> grid(n);
    for (int i = 0; i < n; i ++)
        cin >> grid[i];
    
    vector<vector<int>> vis(n, vector<int>(m));
    vector<vector<pair<int, int>>> pre(n, vector<pair<int, int>>(m));

    int endx = -1, endy = -1;

    vector<int> dirs_u = {1, 0, -1, 0};
    vector<int> dirs_v = {0, 1, 0, -1};

    auto dfs = [&] (auto&& dfs, int u, int v) -> int {
        if (grid[u][v] == 'F')
            endx = u, endy = v;
        for (int i = 0; i < 4; i ++) {
            int du = dirs_u[i], dv = dirs_v[i];
            if ((u + du >= 0) && (u + du < n) && (v + dv >= 0) && (v + dv < m) &&
                (grid[u+du][v+dv] != '*') && !vis[u+du][v+dv]) {
                    pre[u+du][v+dv] = {u, v};
                    vis[u+du][v+dv] = 1;
                    dfs(dfs, u+du, v+dv);
                }
        }
        return 0;
    };

    vis[0][0] = 1;
    dfs(dfs, 0, 0);
    vector<pair<int, int>> path;
    int x = endx, y = endy;

    while (x || y) {
        path.push_back({x, y});
        int nx = pre[x][y].first, ny = pre[x][y].second;
        x = nx, y = ny;
    }

    path.push_back({0, 0});
    reverse(path.begin(), path.end());

    string ud = "UD", lr = "LR";
    bool f_ud = false, f_lr = false;
    int k = path.size();

    for (int i = 1; i < k; i ++) {
        int dx = path[i].first - path[i-1].first, dy = path[i].second - path[i-1].second;
        if (dx == 0) {
            if (!f_lr) {
                f_lr = true;
                int y = op('L').second;
                if (y == path[i].second) lr = "RL";
                else op('R');
            }
            else op(lr[dy > 0]);
        }
        else {
            if (!f_ud) {
                f_ud = true;
                int x = op('U').first;
                if (x == path[i].first) ud = "DU";
                else op('D');
            }
            else op(ud[dx > 0]);
        }
    }

    return 0;
}

File: daily_problems/2024/06/0615/personal_submission/cf811d_yawn_sean.py
# Submission link: https://codeforces.com/contest/811/submission/265821578
def main():
    n, m = MII()
    grid = [I() for _ in range(n)]

    vis = [[0] * m for _ in range(n)]
    vis[0][0] = 1

    prev = [[None] * m for _ in range(n)]

    dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    endx, endy = -1, -1

    stack = [(0, 0)]
    while stack:
        x, y = stack.pop()
        if grid[x][y] == 'F':
            endx, endy = x, y
        for dx, dy in dirs:
            if 0 <= x + dx < n and 0 <= y + dy < m and grid[x+dx][y+dy] != '*' and vis[x+dx][y+dy] == 0:
                vis[x+dx][y+dy] = 1
                prev[x+dx][y+dy] = (x, y)
                stack.append((x+dx, y+dy))

    steps = [(endx, endy)]

    while True:
        x, y = steps[-1]
        if prev[x][y] is None: break
        steps.append(prev[x][y])

    steps.reverse()

    ud = 'UD'
    lr = 'LR'

    f_ud = False
    f_lr = False

    def op(c):
        print(c, flush=True)
        return GMI()

    k = len(steps)
    for i in range(k - 1):
        dx, dy = steps[i+1][0] - steps[i][0], steps[i+1][1] - steps[i][1]
        if dx == 0:
            if not f_lr:
                f_lr = True
                nx, ny = op('L')
                if ny == steps[i+1][1]: lr = 'RL'
                else: op('R')
            else:
                if dy > 0: op(lr[1])
                else: op(lr[0])
        else:
            if not f_ud:
                f_ud = True
                nx, ny = op('U')
                if nx == steps[i+1][0]: ud = 'DU'
                else: op('D')
            else:
                if dx > 0: op(ud[1])
                else: op(ud[0])